### DISCUSSION TITLE
IDE not recognizing firstOrCreate

### DISCUSSION THREAD

**Discussion by MikeMazalahi** [spatie/laravel-permission#1728] [Q&A] [answered]
Hello,
I've been using PHPStorm to develop Laravel websites, I have a class in which I'm seeding my permissions.
To account for multiple seedings not crashing, I'm creating my roles and permissions using Eloquent's `firstOrCreate`method as such:

```php
Permission::firstOrCreate(['name' => 'create user']);
$admin = Role::firstOrCreate(['name' => 'admin']);
$admin->syncPermissions('create user');
```

This works OK in code (which is why I haven't filed it as a bug) but when evaluating through PHPStorm, I get these warnings:

```
Method 'firstOrCreate' not found in \Spatie\Permission\Models\Permission
Method 'firstOrCreate' not found in \Spatie\Permission\Models\Role
```

This issue is present with my own models as well, solved through mixing in the Eloquent Builder docs with my class phpdoc:

```php
use Illuminate\Database\Eloquent\Builder;
/**
 * Class User
 *
 * @package App\Models
 * @mixin Builder
 */
```

The question now is, do I have a way of solving this for my Role and Permission models without having to re-implement many of the features that are provided in the default model? and are there any issues with adding that mixin in the default models? I would create the pull request if that's allowed.

**Comment by MikeMazalahi** [comment]
I'm sorry for asking for a direct review, but since it has been a week, @drbyte could you perhaps give me an insight on your policy for this? Cloning the repository, I realized there isn't a phpdoc block in the file at all. Considering this, should I work on this issue and create a pull request on it?

**Reply by drbyte** [reply]
Spatie's [guidelines](https://spatie.be/guidelines/laravel-php#docblocks) do cover docblocks for the situations where you do choose/need to add them.

That said, this package, and in the other most popular Spatie packages there are very few docblocks, and I find only one mention of a mixin which is for an internal class not for a Laravel class.

In this package we've opted to use few docblocks except where they're needed for specific clarity.

**Comment by drbyte** [comment, marked as answer]
Extending the model and adding the docblock and mixin notation to it will also work, without having to re-implement the class.

**Reply by MikeMazalahi** [reply]
That will resolve the issue I'm having, Thanks! Although I have to ask, do you (or Spatie) have a policy on phpdocs? (or a lack thereof)
